onnx/defs/tensor/defs.cc

```diff
@@ -242,37 +242,22 @@ will be (2, 1, 3).
         {"tensor(float16)", "tensor(float)", "tensor(double)"},
         "Constrain input and output types to float tensors.")
     .ShapeInferenceFunction([](InferenceContext& ctx) {
-      if (ctx.getNumInputTypes() != 1) {
-        return;
-      }
-      std::vector<int64_t> perm;
-      {
-        auto perm_attr = ctx.getAttribute("perm");
-        if (perm_attr) {
-          for (auto dim : perm_attr->ints()) {
-            perm.push_back(dim);
-          }
-        } else {
-          int ndims = ctx.getInputType(0)->shape().dim_size();
-          for (int i = 0; i < ndims; i++) {
-            perm.push_back(ndims - i - 1);
-          }
+        if (ctx.getNumInputTypes() != 1 || !ctx.getInputType(0)) {
+          return;
         }
-      }
 
-      TypeProto_Tensor tt;
-      {
-        tt.set_elem_type(ctx.getInputType(0)->elem_type());
-        auto shape = tt.mutable_shape();
+        std::vector<int64_t> perm;
+        if (!getRepeatedAttribute(ctx, "perm", perm)) {
+          for (int i = ctx.getInputType(0)->shape().dim_size() - 1; i >= 0; i--) {
```

Comment:
so what would be the shape if we don't even know the rank of some tensor? would you expect that shape won't be set at all in this case? Then we need to test for it somehow

Reply:
this is implementing "By default, reverse the dimensions" from the ONNX spec for transpose. I can add a comment

Reply:
ah I see what you mean. I guess there are two cases

- there is no type information at all (handled at the top of the function). This represents unknown rank

- dim_size() == 0. This would represent a scalar tensor, and this code should handle that fine (it puts nothing in the output type dimensions). I'll add a test for this case